feat: add docs for authMiddleware() (#162)
clerkinc/clerk-docs: docs/references/nextjs/auth-middleware.mdx

---
title: authMiddleware() | Next.js
description: The `authMiddleware()` method allows you to protect your Next.js application using middleware.
---

# `authMiddleware()`

The `authMiddleware()` helper integrates Clerk authentication into your Next.js application through middleware. `authMiddleware()` is compatible with both the App and Pages routers.

## Usage

```ts filename="middleware.ts" copy
import { authMiddleware } from "@clerk/nextjs";

export default authMiddleware();

export const config = {
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

### Using `afterAuth()` for fine-grained control

> Some developers will need to handle specific cases such as detecting if the user is inside an organization, or handle redirects differently. This cases can be handled with `afterAuth()`.

Reviewer: ```suggestion
Some developers will need to handle specific cases such as handling redirects differently or detecting if a user is inside an organization. These cases can be handled with `afterAuth()`.
```

```ts filename="middleware.ts" copy
import { authMiddleware, redirectToSignIn } from '@clerk/nextjs';

export default authMiddleware({
  afterAuth(auth, req, evt) {
    // handle users who aren't authenticated
    if (!auth.userId && !auth.isPublicRoute) {
      return redirectToSignIn({ returnBackUrl: req.url });
    }
    // redirect them to organization selection page
    if(auth.userId && !auth.orgId && req.nextUrl.pathname !== "/org-selection"){
      const orgSelection = new URL('/org-selection', req.url)
      return NextResponse.redirect(orgSelection)
    }
  }
});
```

### Using `beforeAuth()` to execute middleware before authentication

> If you need to additional middleware handlers to execute before Clerk's authentication middleware, use `beforeAuth()`. An example where the middleware handler from `next-intl` is executed can be seen below.

Reviewer: ```suggestion
If you need additional middleware handlers to execute before Clerk's authentication middleware, use `beforeAuth()`. An example where the middleware handler from `next-intl` is executed can be seen below.
```

```ts filename="middleware.ts" copy
import { authMiddleware } from "@clerk/nextjs";

import createMiddleware from "next-intl/middleware";

const intlMiddleware = createMiddleware({
  locales: ["en", "el"],

  defaultLocale: "en",
});

export default authMiddleware({
  beforeAuth: (req) => {
    // Execute next-intl middleware before Clerk's auth middleware
    return intlMiddleware(req);
  },

  // Ensure that locale specific sign-in pages are public
  publicRoutes: ["/", "/:locale/sign-in"],
});

export const config = {
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

### Making pages public using `publicRoutes`

By default, Clerk's `authMiddleware()` treats all routes as private if the middleware runs. If you need to make specific routes public, use the `publicRoutes` option.

```ts filename="middleware.ts" copy
import { authMiddleware } from "@clerk/nextjs";
export default authMiddleware({
  // "/" will be accessible to all users
  publicRoutes: ["/"]
});

export const config = {
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

> If you are building your own sign in pages, you don't need to add these pages to your publicRoutes. They will be inferred from your `.env` file:

Reviewer: ```suggestion
If you are building your own sign in pages, you don't need to add these pages to your `publicRoutes`. They will be inferred from your `.env` file:
```
